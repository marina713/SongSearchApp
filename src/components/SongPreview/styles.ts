import styled from 'styled-components/native';
import {ThemeColors} from '~/constants/colors';

export const Container = styled.TouchableOpacity<{theme: ThemeColors}>`
  margin: 5px;
  border-radius: 15px;
  border-color: ${({theme}) => theme.border};
  border-width: 1px;
  padding: 8px 12px;
  flex-direction: row;
`;

export const FlexCol = styled.View`
  flex-direction: column;
  flex: 1;
`;

export const FlexRow = styled.View`
  flex-direction: row;
  flex-wrap: wrap;
`;

export const ExtraInfoContainer = styled.View<{theme: ThemeColors}>`
  flex-direction: row;
  justify-content: center;
  border: 0.7px solid ${({theme}) => theme.text};
  border-radius: 20px;
  padding: 2px 10px;
  margin-top: 5px;
`;

export const TrackName = styled.Text<{theme: ThemeColors}>`
  font-size: 22px;
  color: ${({theme}) => theme.title};
  font-weight: 600;
`;
export const AlbumCover = styled.Image`
  aspect-ratio: 1;
  width: 85px;
  border-radius: 8px;
  margin: 5px 15px 5px 0;
`;

export const Title = styled.Text`
  font-size: 17px;
  color: ${({theme}) => theme.title};
  margin: 3px 0;
`;
export const ExtraInfoText = styled.Text<{theme: ThemeColors}>`
  font-size: 16px;
  color: ${({theme}) => theme.text};
`;
export const SubTitle = styled.Text<{theme: ThemeColors}>`
  font-size: 13px;
  color: ${({theme}) => theme.subtitle};
  margin: 1px 0;
`;
