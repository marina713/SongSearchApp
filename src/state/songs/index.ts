import {
  UPDATE_SONGS,
  initialState,
  SET_SORT_BY,
  SET_PLAYING_TRACK_ID,
} from './constants';
import {SongsState, Action} from './types';

export const songs = (
  state: SongsState = initialState,
  action: Action,
): SongsState => {
  switch (action.type) {
    case UPDATE_SONGS: {
      const {payload} = action;
      return {
        ...state,
        songs: payload,
        sortBy: initialState.sortBy,
      };
    }

    case SET_SORT_BY: {
      const {payload} = action;
      return {
        ...state,
        sortBy: payload,
      };
    }

    case SET_PLAYING_TRACK_ID: {
      const {payload} = action;
      return {
        ...state,
        playingTrackId: payload,
      };
    }

    default: {
      return state;
    }
  }
};
