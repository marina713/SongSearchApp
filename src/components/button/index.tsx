import * as React from 'react';

import { Container, ButtonText } from './styles';

export type Props = {
  text: string;
  onPress: () => void;
};

export const Button = React.memo(({ text, onPress }: Props) => (
  <Container onPress={onPress}>
    <ButtonText>{text}</ButtonText>
  </Container>
));
