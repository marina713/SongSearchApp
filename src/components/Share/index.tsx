import React from 'react';
import { Share, Image, TouchableOpacity } from 'react-native';
import { Song } from '~/state/songs/types';
import { Container } from './styles';

type Props = { song: Song };

export const ShareButton = ({ song }: Props) => {
  const onShare = async () => {
    try {
      const result = await Share.share({
        message: `Check out this song: ${song.trackName} by ${song.artistName}`,
      });
      if (result.action === Share.sharedAction) {
        if (result.activityType) {
          // shared with activity type of result.activityType
        } else {
          // shared
        }
      } else if (result.action === Share.dismissedAction) {
        // dismissed
      }
    } catch (error) {}
  };
  return (
    <Container>
      <TouchableOpacity onPress={onShare}>
        <Image
          source={require('~/assets/images/share.png')}
          style={{ width: 40, height: 40, margin: 5 }}
        />
      </TouchableOpacity>
    </Container>
  );
};
