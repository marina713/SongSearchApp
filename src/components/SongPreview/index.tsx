import * as React from 'react';
import { Song } from '~/state/songs/types';
import { msToHMS, formatDate } from '~/utils/time';

import {
  Container,
  TrackName,
  Title,
  AlbumCover,
  FlexCol,
  FlexRow,
  ExtraInfoContainer,
  ExtraInfoText,
  SubTitle,
} from './styles';

export type Props = {
  song: Song;
  onSongPress?: () => void;
};

export const SongPreview = React.memo(({ song, onSongPress }: Props) => {
  const price =
    song.currency === 'USD'
      ? `$${song.trackPrice}`
      : `${song.trackPrice}${song.currency}`;
  const formattedReleaseDate = song.releaseDate && formatDate(song.releaseDate);
  const releaseDate = formattedReleaseDate
    ? `Released on ${formattedReleaseDate} · `
    : '';
  const formattedLength = song.trackTimeMillis && msToHMS(song.trackTimeMillis);
  const length = formattedLength ? `Length ${formattedLength}` : '';

  return (
    <Container onPress={onSongPress}>
      <AlbumCover source={{ uri: song.artworkUrl100 }} />
      <FlexCol>
        <TrackName>{song.trackName}</TrackName>
        <Title>{`${song.artistName} · ${song.primaryGenreName}`}</Title>
        <SubTitle>{`${releaseDate}${length}`}</SubTitle>
        <FlexRow>
          <ExtraInfoContainer>
            <ExtraInfoText>{price}</ExtraInfoText>
          </ExtraInfoContainer>
        </FlexRow>
      </FlexCol>
    </Container>
  );
});
