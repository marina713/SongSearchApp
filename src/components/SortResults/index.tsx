import * as React from 'react';
import { setSortBy } from '~/state/songs/actions';
import { useSelector, useDispatch } from 'react-redux';
import { getSortBy } from '~/state/songs/selectors'

import { Container, Title, SortItem, SortItemText, SortItemsContainer, SortIcon } from './styles';
import { SortByType } from '~/state/songs/types';

type SortItemComponentProps = {
    sortBy: SortByType;
}

const SORT_BY = ['Duration', 'Genre', 'Price'];

const SortItemComponent = ({ sortBy }: SortItemComponentProps) => {
    const dispatch = useDispatch();
    const sortByState = useSelector(getSortBy);

    const isSelected = sortByState.type === sortBy;
    const source = sortByState.order === 'DESC' ?
        require('~/assets/images/sort-descending.png')
        : require('~/assets/images/sort-ascending.png');

    const onPress = () => {
        const nextValue = sortByState.type === '' || !isSelected ?
            { type: sortBy, order: 'DESC' }
            : sortByState.order === 'DESC' ?
                { type: sortBy, order: 'ASC' } : { type: '', order: '' };
        dispatch(setSortBy(nextValue))
    }

    return (
        <SortItem isSelected={isSelected} onPress={onPress}>
            <SortItemText isSelected={isSelected}>{sortBy}</SortItemText>
            {isSelected ? <SortIcon source={source} /> : null}
        </SortItem>
    )
}

export const SortResults = () => (
    <Container>
        <Title>Sort by</Title>
        <SortItemsContainer>
            {SORT_BY.map(item => <SortItemComponent key={item} sortBy={item} />)}
        </SortItemsContainer>
    </Container>
);
