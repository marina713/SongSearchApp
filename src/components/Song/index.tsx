import * as React from 'react';
import { Song as SongType } from '~/state/songs/types';
import { msToHMS, formatDate } from '~/utils/time'
import { ShareButton } from '../Share';

import {
    Container,
    TrackName,
    Title,
    AlbumCover,
    FlexRow,
    ExtraInfoContainer,
    ExtraInfoText,
    FlexRowContainer
} from './styles';

export type Props = {
    song: SongType;
};

export const Song = React.memo(({ song }: Props) => {
    const formattedReleaseDate = song.releaseDate && `Released ${formatDate(song.releaseDate)}`;
    const formattedLength = song.trackTimeMillis && `Length ${msToHMS(song.trackTimeMillis)}`;
    const price = song.currency === 'USD' ? `$${song.trackPrice}` : `${song.trackPrice}${song.currency}`;
    const genre = song.primaryGenreName || '';

    const extraInfo = [genre, formattedReleaseDate, formattedLength, price];

    return (
        <Container>
            <TrackName>{song.trackName}</TrackName>
            <Title>{song.artistName}</Title>
            <AlbumCover source={{ uri: song.artworkUrl100 }} />
            <FlexRowContainer>
                <ShareButton song={song} />
                <FlexRow>
                    {extraInfo.map(item => (item ?
                        <ExtraInfoContainer key={item}>
                            <ExtraInfoText>{item}</ExtraInfoText>
                        </ExtraInfoContainer> : null))}
                </FlexRow>
            </FlexRowContainer>
        </Container>
    )
});
