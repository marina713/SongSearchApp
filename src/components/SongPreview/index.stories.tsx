import { action } from '@storybook/addon-actions';
import { text, number } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import React from 'react';

import { SongPreview } from '~/components/SongPreview';
import { CenterView } from '~/components/container/center-view';

storiesOf('SongPreview', module)
  .addDecorator(getStory => <CenterView>{getStory()}</CenterView>)
  .add('with song', () => (
    <SongPreview
      onSongPress={action('clicked-song')}
      song={{
        artistId: 156488786,
        trackId: 1444896574,
        artistName: text('Artist Name', 'Sam Smith'),
        trackName: text('Track Name', 'Nirvana'),
        trackPrice: number('Price', 0.62),
        releaseDate: '2013-01-01T12:00:00Z',
        trackTimeMillis: 197551,
        currency: 'USD',
        primaryGenreName: 'Pop',
        artworkUrl100:
          'https://is4-ssl.mzstatic.com/image/thumb/Music114/v4/e4/f9/59/e4f959ff-e758-ff29-4b26-39ae222489c8/source/100x100bb.jpg',
      }}
    />
  ));
