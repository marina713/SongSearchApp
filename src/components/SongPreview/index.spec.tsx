import * as React from 'react';
import renderer from 'react-test-renderer';
import { MOCK_SONGS } from '~/state/songs/mocks';

import { SongPreview, Props } from '~/components/SongPreview';

jest.mock('./styles', () => ({
  Container: 'Container',
  TrackName: 'TrackName',
  Title: 'Title',
  AlbumCover: 'AlbumCover',
  ExtraInfoContainer: 'ExtraInfoContainer',
  ExtraInfoText: 'ExtraInfoText',
  FlexRow: 'FlexRow',
  FlexCol: 'FlexCol',
  SubTitle: 'SubTitle'
}));

const initialProps = {
  song: MOCK_SONGS[0],
  onSongPress: () => { }
};

const testComponent = (props?: Props) => {
  return renderer.create(<SongPreview {...initialProps} {...props} />);
};

describe('<SongPreview />', () => {
  it('renders correctly with default props', () => {
    const tree = testComponent().toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with custom Song', () => {
    const tree = testComponent({
      ...initialProps,
      song: MOCK_SONGS[2],
    }).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
