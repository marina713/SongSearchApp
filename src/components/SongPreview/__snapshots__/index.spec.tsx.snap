// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SongPreview /> renders correctly with custom Song 1`] = `
<Container
  onPress={[Function]}
>
  <AlbumCover
    source={
      Object {
        "uri": "https://is3-ssl.mzstatic.com/image/thumb/Music125/v4/8d/58/a9/8d58a928-0663-71f7-4e5c-d2bd54a124fb/source/100x100bb.jpg",
      }
    }
  />
  <FlexCol>
    <TrackName>
      Come As You Are
    </TrackName>
    <Title>
      Nirvana 路 Alternative
    </Title>
    <SubTitle>
      Released on 24/9/1991 路 Length 3:39
    </SubTitle>
    <FlexRow>
      <ExtraInfoContainer>
        <ExtraInfoText>
          $1.29
        </ExtraInfoText>
      </ExtraInfoContainer>
    </FlexRow>
  </FlexCol>
</Container>
`;

exports[`<SongPreview /> renders correctly with default props 1`] = `
<Container
  onPress={[Function]}
>
  <AlbumCover
    source={
      Object {
        "uri": "https://is4-ssl.mzstatic.com/image/thumb/Music114/v4/e4/f9/59/e4f959ff-e758-ff29-4b26-39ae222489c8/source/100x100bb.jpg",
      }
    }
  />
  <FlexCol>
    <TrackName>
      Nirvana
    </TrackName>
    <Title>
      Sam Smith 路 Pop
    </Title>
    <SubTitle>
      Released on 1/1/2013 路 Length 3:17
    </SubTitle>
    <FlexRow>
      <ExtraInfoContainer>
        <ExtraInfoText>
          $0.69
        </ExtraInfoText>
      </ExtraInfoContainer>
    </FlexRow>
  </FlexCol>
</Container>
`;
